File: wheel/src/main/java/user11681/wheel/util/GroovyUtil.java
package user11681.wheel.util;

import java.util.Arrays;
import java.util.List;
import org.codehaus.groovy.runtime.callsite.CallSite;
import user11681.reflect.Invoker;
import user11681.uncheck.Uncheck;

public class GroovyUtil {
    public static List<CallSite> sites(Class<?> type) {
        return Arrays.asList(Invoker.invoke(Invoker.findStatic(type, "$getCallSiteArray", CallSite[].class)));
    }

    public static CallSite site(Class<?> type, int index) {
        return sites(type).get(index);
    }

    public static Object property(List<CallSite> sites, Object object, int... indexes) {
        for (int index : indexes) {
            Object javaBad = object;
            object = Uncheck.handle(() -> sites.get(index).callGetProperty(javaBad));
        }

        return object;
    }

    public static Object call(List<CallSite> sites, Object object, int... indexes) {
        for (int index : indexes) {
            Object javaBad = object;
            object = Uncheck.handle(() -> sites.get(index).call(javaBad));
        }

        return object;
    }
}

File: src/net/auoeke/wheel/util/GroovyUtil.kt
package net.auoeke.wheel.util

import net.auoeke.extensions.type
import net.auoeke.reflect.Invoker
import org.codehaus.groovy.runtime.callsite.CallSite

@Suppress("NAME_SHADOWING")
object GroovyUtil {
    private fun sites(type: Class<*>?): List<CallSite> {
        return listOf(*Invoker.invoke(Invoker.findStatic(type, "\$getCallSiteArray", type<Array<CallSite>>())))
    }

    fun site(type: Class<*>?, index: Int): CallSite {
        return this.sites(type)[index]
    }

    fun property(sites: List<CallSite>, o: Any?, vararg indexes: Int): Any? {
        var o = o

        indexes.forEach {
            o = sites[it].callGetProperty(o)
        }

        return o
    }

    fun call(sites: List<CallSite>, o: Any?, vararg indexes: Int): Any? {
        var o = o

        indexes.forEach {
            o = sites[it].call(o)
        }

        return o
    }
}
